src/analyses/ai.h

```diff
@@ -96,6 +104,29 @@ class ai_baset
     finalize();
   }
 
+  /// Returns all of the histories that have reached
+  /// the start of the instruction.
+  /// PRECONDITION(l is dereferenceable)
+  /// \param l: The location before which we want the set of histories
+  /// \return The set of histories before `l`.
+  virtual ctrace_set_ptrt abstract_traces_before(locationt l) const
+  {
+    return storage->abstract_traces_before(l);
+  }
+
+  /// Returns all of the histories that have reached
+  /// the end of the instruction.
+  /// PRECONDITION(l is dereferenceable)
+  /// \param l: The location before which we want the set of histories
+  /// \return The set of histories before `l`.
+  virtual ctrace_set_ptrt abstract_traces_after(locationt l) const
+  {
+    /// PRECONDITION(l is dereferenceable && std::next(l) is dereferenceable)
+    /// Check relies on a DATA_INVARIANT of goto_programs
+    INVARIANT(!l->is_end_function(), "No state after the last instruction");
+    return storage->abstract_traces_before(l);
```

Comment:
The precondition suggests we should pass `std::next(l)` here. Just checking this is intentional.

Reply:
Nope; you are correct and that is a bug.  Fixed.  Documentation and preconditions, even not enforced ones, are useful!